### PR TITLE
Chore: add test case for empty file #6

### PR DESCRIPTION
**Pull request by niedzielski**
Marked pending since functionality is not yet available.

### CODE REVIEW

File: test/css-validator_test.js [hunk 1]
```diff
@@ -19,6 +19,12 @@ function runValidateCss() {
   });
 }
 
+function expectValid(data) {
+  expect(data.validity).to.equal(true);
+  expect(data.errors).to.deep.equal([]);
+  expect(data.warnings).to.deep.equal([]);
+}
```
Comment (twolfson) [inline test/css-validator_test.js]: This is a personal preference but I hate consolidated assertions. It loses clarity of what's being asserted inside of the test case which can introduce bugs. Can we please use the long form in both locations?

**Comment by twolfson** [comment]
This is a good start but it looks like it's not reproducing the error in #6. Can we correct the test so it does?

**Comment by niedzielski** [comment]
@twolfson, I think you're referring to my usage of Mocha skip() on the test? I prefer skipping failing tests since failing tests make it difficult to make unrelated changes and they still show in the report as "pending". I've removed the call to skip() :]

**Comment by twolfson** [comment]
Ah I totally missed the skip the first time. The tests look good now. Can
we add a patch to the code as well in this PR?

I prefer to land patches and tests in the same PR to prevent logical
discontinuity (eg forget to test an edge case)

File: lib/validation-stream.js [hunk 2]
```diff
@@ -9,6 +9,12 @@ var Duplex = require('readable-stream').Duplex;
 
 var XmlParser = require('./xml-parser');
 
+/** @arg {?string} str Input string to check.
+    @return {!boolean} True if input is undefined, null, or entirely whitespace. */
```
Comment (twolfson) [inline lib/validation-stream.js]: For consistency with the rest of the repo, we should not add any JSDoc or similar (as the other functions don't have it)

File: lib/validation-stream.js [hunk 3]
```diff
@@ -9,6 +9,12 @@ var Duplex = require('readable-stream').Duplex;
...
+function isBlank(str) {
+  return !str || !/\S/.test(str);
```
Comment (twolfson) [inline lib/validation-stream.js]: `!str` seems redundant to the regexp check =/

File: lib/validation-stream.js [hunk 4]
```diff
@@ -9,6 +9,12 @@ var Duplex = require('readable-stream').Duplex;
 
 var XmlParser = require('./xml-parser');
 
+/** @arg {?string} str Input string to check.
+    @return {!boolean} True if input is undefined, null, or entirely whitespace. */
+function isBlank(str) {
```
Comment (twolfson) [inline lib/validation-stream.js]: We shouldn't need to define this as a helper function as it's only being used once. My preference is to only break out helpers when it's used at least twice/thrice or we will need it at least twice/thrice.

Instead, we can define this boolean directly in the constructor:

```js
var isBlank = /\S/.test(str);
if (isBlank) { /* ... */ }
```

File: lib/validation-stream.js [hunk 5]
```diff
@@ -18,11 +24,17 @@ function ValidationStream(options) {
   // Create internal placeholder for CSS
   this._css = [];
 
-  // Generate our request
-  this.generateRequest();
+  if (isBlank(this.options.text)) {
+    // The W3C service only understands nonempty CSS. Fork a dummy event to be acted on after the
+    // client has yielded.
+    setTimeout(this._emitEmptyCssEvents.bind(this), 0);
```
Comment (twolfson) [inline lib/validation-stream.js]: Great thinking here! In Node.js, we can use something even faster than `setTimeout` yet async -- `process.nextTick`.

Also, I always like to document this async action with a comment about preventing zalgo for future devs

File: lib/validation-stream.js [hunk 6]
```diff
@@ -160,6 +172,15 @@ extend(ValidationStream.prototype, {
...
     // DEV: Do not take any action as we only emit one data event, the result
+  },
+  _emitEmptyCssEvents: function () {
+    var data = {
+      validity: true,
+      errors: [],
+      warnings: []
+    };
+    this.emit('data', data);
+    this.emit('end');
```
Comment (twolfson) [inline lib/validation-stream.js]: While streams are event emitters, we should stick to their built-in API. We should use `push(result)` and `push(null)`:

https://github.com/twolfson/css-validator/blob/0.7.0/lib/validation-stream.js#L150-L152

https://nodejs.org/api/stream.html#stream_readable_push_chunk_encoding

**Comment by twolfson** [comment]
Looks good, mostly nitpicking on coding style/arrangement. Going to do the same commandeering business again

**Comment by twolfson** [comment]
One more issue that came up during comandeering -- we didn't handle blank check for URIs. I have added that in now and will add a test as well

**Comment by twolfson** [comment]
It turns out we never tested URI support which lead to a whole bunch of refactoring. Everything is green now which is nice :+1: This has been landed/released in `0.8.0`. Thanks for the PR!

File: lib/validation-stream.js (hunk 4 above)
Comment (niedzielski) [inline lib/validation-stream.js]: Although I didn't unit test this function, I prefer many small functions for unit testing as well as composability and because I don't want to think about more than I have to at any time. One of the downsides is that the whole is harder to see.

File: lib/validation-stream.js (hunk 3 above)
Comment (niedzielski) [inline lib/validation-stream.js]: isBlank() returns true if the input is undefined, null, or entirely whitespace. Here are some of the test cases I should have added: `isBlank(null)` => true, `isBlank(undefined)` => true, `isBlank(' ')` => true, `isBlank('A')` => false. When I remove the leading `!str ||`, the first two cases fail.

File: lib/validation-stream.js (hunk 2 above)
Comment (niedzielski) [inline lib/validation-stream.js]: I find that annotating nullity expectations is invaluable but to each their own. I would recommend at least keeping a comment about the return value since different folks might have different expectations over what "blank" means. I used the definition from [Apache Commons' StringUtils](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#isBlank-java.lang.CharSequence-).

File: lib/validation-stream.js (hunk 5 above)
Comment (niedzielski) [inline lib/validation-stream.js]: I really appreciate you taking the time to share this tip! I'm still ramping up on Node.js (and JavaScript in general) so your feedback is dear!

File: lib/validation-stream.js (hunk 6 above)
Comment (niedzielski) [inline lib/validation-stream.js]: Thanks!

**Comment by niedzielski** [comment]
@twolfson, thank you!

File: lib/validation-stream.js (hunk 4 above)
Comment (twolfson) [inline lib/validation-stream.js]: Yep, I typically prefer to only break out code when reuse is necessary or the function is getting too big and state is getting hard to track/manage.

One downside to unit testing for something like this is it's typically overkill as you would need to not only test the method itself but also have some spy-esque assertion to verify we are using `isBlank` in the code, at the appropriate time, with the expected content (aka a contract).

File: lib/validation-stream.js (hunk 2 above)
Comment (twolfson) [inline lib/validation-stream.js]: I'm not bashing the documentation itself. This was mostly a comment on keeping the code base consistent. If we want to document detecting nullity and such, then we could have added an inline comment:

```js
function isBlank(str) {
  // Returns true if input is undefined, null, or entirely whitespace
  return !str || ...;
}
```

File: lib/validation-stream.js (hunk 3 above)
Comment (twolfson) [inline lib/validation-stream.js]: Ah, I totally missed that. The name `str` implies that we expect a string and never `undefined` or `null`. A better name might have been `input`. Additionally, we could make it clearer what the code is trying to achieve by doing something like:

```js
return typeof input === 'string' ? !/\S/.test(str) : false;
```
